api/src/main/java/com/epam/esm/web/controller/CertificateController.java

```diff
@@ -0,0 +1,170 @@
+            @RequestParam(required = false, name = "tag") String tag,
+            @RequestParam(required = false, name = "sort_by") String sortObject,
+            @RequestParam(required = false, name = "order_by") String sortType,
+            @RequestParam(required = false, name = "search_by") String searchType,
+            @RequestParam(required = false, name = "search_value") String searchValue) {
+        List<GiftCertificateDto> certificates;
+        if (searchType != null) {
+            certificates = searchByParams(searchType, searchValue);
+        } else {
+            certificates = service.readAll();
+        }
+        if (tag != null) {
+            List<GiftCertificateDto> certificatesByTag = service.searchByTagName(tag);
+            certificates.retainAll(certificatesByTag);
+        }
+        processCertificatesSort(certificates, sortObject, sortType);
+        return certificates;
+    }
+
+    /**
+     * Gets {@link GiftCertificateDto} by id
+     *
+     * @param id {@link GiftCertificateDto} id
+     * @return {@link GiftCertificateDto} with given id
+     */
+    @GetMapping(value = "/{id}", produces = "application/json")
+    @ResponseStatus(HttpStatus.OK)
+    public GiftCertificateDto receiveGiftCertificate(@PathVariable int id) {
+        return service.read(id);
+    }
+
+    /**
+     * Creates new gift certificate
+     *
+     * @param giftCertificate {@link GiftCertificateDto} with create parameters
+     * @return newly created {@link GiftCertificateDto}
+     */
+    @PostMapping(consumes = "application/json", produces = "application/json")
+    @ResponseStatus(HttpStatus.CREATED)
+    public GiftCertificateDto createGiftCertificate(@Valid @RequestBody GiftCertificateDto giftCertificate) {
+        return service.create(giftCertificate);
+    }
+
+    /**
+     * Deletes gift certificate by given id
+     *
+     * @param id gift certificate id
+     * @return server response
+     */
+    @DeleteMapping(value = "/{id}")
+    public ResponseEntity<?> deleteGiftCertificate(@PathVariable int id) {
+        service.delete(id);
+        return ResponseEntity.noContent().build();
+    }
+
+    /**
+     * Updates gift certificate
+     *
+     * @param id              gift certificate id
+     * @param giftCertificate {@link GiftCertificateDto} object with updated parameters
+     * @return updated {@link GiftCertificateDto}
+     */
+    @PutMapping(value = "/{id}", consumes = "application/json", produces = "application/json")
+    @ResponseStatus(HttpStatus.OK)
+    public GiftCertificateDto updateGiftCertificate(@PathVariable int id,
+                                                    @RequestBody GiftCertificateDto giftCertificate) {
+        giftCertificate.setId(id);
+        return service.update(giftCertificate);
+    }
+
+    private List<GiftCertificateDto> searchByParams(String searchType, String searchValue) {
+        List<GiftCertificateDto> certificates;
+        switch (searchType) {
```

Comment:
It is possible to serarch together:
- by tag and name 
- by tag and description
Search by 3 params together was not implemented.
As I just noticed, "tag" case branch should be removed.

Reply:
Hm, passing `searchType` as parameters seems strange to me since you can make properties not required and check for nulls, this will make API little easier

Reply:
It really makes sense. Thanks for the advice.